pennylane/devices/qubit/sampling.py

```diff
@@ -42,18 +44,19 @@ def measure_with_samples(
     for op in mp.diagonalizing_gates():
         pre_rotated_state = apply_operation(op, pre_rotated_state)
 
-    # we don't need to worry about shot vectors for now
-    # if shots.has_partitioned_shots:
-    #     processed_samples = []
-    #     for shot_copies in shots.shot_vector:
-    #         for _ in range(shot_copies.copies):
-    #             samples = sample_state(pre_rotated_state, shot_copies.shots, rng=rng)
-    #             processed_samples.append(mp.process_samples(samples, wire_order))
+    wires = qml.wires.Wires(range(len(state.shape)))
 
-    #     return tuple(processed_samples)
+    # if there is a shot vector, build a list containing results for each shot entry
+    if shots.has_partitioned_shots:
+        processed_samples = []
+        for s in shots:
+            samples = sample_state(pre_rotated_state, shots=s, wires=wires, rng=rng)
+            processed_samples.append(mp.process_samples(samples, wires))
```

Comment:
We have two potential ways to do this logic:

1) for each number of shots, sample and post-processing the samples
2) sample `total_shots` times, and then select out which ones we want to post-process.

Current `QubitDevice` and `DefaultQubit` logic does option 2. This was assumed to be faster. Let's actually write up both options and numerically determine which one is better with benchmarking.

Reply:
Option 2 is almost certainly faster but requires more space. When writing this code I kept in mind the reason we use sparse shot vectors in the first place, which is because they use less memory. If, for example, the user asks for 1000 copies of 1000 shots, then instead of producing 1 million samples all at once, we only produce 1000 samples at a time.

Thoughts?

Reply:
Thinking about it again, we should bypass these kinds of optimizations for the moment.

Right now, we just want to get the behaviour unblocked. We can come back and make all kinds of optimizations later on. Let's leave it as is.

Reply:
Alright, I'll add a code comment on it.

Reply:
If this is something we will definitely come back to later a Shortcut story is a good idea as well.